pkg/standalone-utils/contracts/StableLPOracle.sol

```diff
@@ -0,0 +1,213 @@
+
+        k = _findInitialGuessForK(a, b, prices);
+        for (uint256 i = 0; i < 255; i++) {
+            // dTdk and dPdk are the derivatives of T and P with respect to k.
+            // solhint-disable-next-line var-name-mixedcase
+            (int256 T, int256 dTdk, int256 dPdkDivP, int256 alphaDivPTn) = _computeKParams(k, a, b, prices);
+
+            int256 fk = T - alphaDivPTn;
+            int256 dFkdk = (int256(_totalTokens) + 1) * dTdk + _mulDownInt(T, dPdkDivP);
+
+            int256 newK = k - _divDownInt(fk, dFkdk);
+
+            if (newK > k) {
+                if (_divDownInt(newK - k, k) <= _K_MAX_ERROR) {
+                    return newK;
+                }
+            } else if (_divDownInt(k - newK, k) <= _K_MAX_ERROR) {
+                return newK;
+            }
+
+            k = newK;
+        }
+
+        revert KDidntConverge();
+    }
+
+    function _findInitialGuessForK(int256 a, int256 b, int256[] memory prices) internal view returns (int256 k) {
+        // The initial guess for K is important, since f(k) has many roots that return negative balances. We need to
+        // choose a guess where the function is convex.
+        // The best initial guess for K is `k = a * min([(1 + 1/(1 + b)), (2 - b/a)]) / min(price)`.
+
+        int256 minPrice = prices[0];
+        for (uint256 i = 1; i < _totalTokens; i++) {
+            if (prices[i] < minPrice) {
+                minPrice = prices[i];
+            }
+        }
+
+        int256 term1 = _ONE_INT + _divDownInt(_ONE_INT, (_ONE_INT + b));
+        int256 term2 = 2 * _ONE_INT - _divDownInt(b, a);
+
+        if (term1 < term2) {
+            return (term1 * a) / minPrice;
+        } else {
+            return (term2 * a) / minPrice;
+        }
+    }
+
+    function _computeKParams(
+        int256 k,
+        int256 a,
+        int256 b,
+        int256[] memory prices
+    ) internal view returns (int256 T, int256 dTdk, int256 dPdkDivP, int256 alphaDivPTn) {
+        // solhint-disable-previous-line var-name-mixedcase
+
+        uint256 i;
+        int256 den;
+
+        // To avoid overflows, we divided f(k) and f'(k) by PT^n. That's why, instead of computing P, P' and alpha,
+        // we change the variables to compute `dPdkDivP = P'/P` and `alphaDivPTn = alpha/(PT^n)`.
+
+        T = 0;
+        // dTdk = T', where T' is the derivative of T with respect to k.
+        dTdk = 0;
+        // dPdkDivP = P'/P, where P' is the derivative of P with respect to k.
+        dPdkDivP = 0;
+        // alphaDivPTn = alpha/(PT^n), where alpha is the constant term of the stable invariant.
+        alphaDivPTn = b;
+        for (i = 0; i < _totalTokens; i++) {
+            den = _mulDownInt(k, prices[i]) - a;
+            T += _divDownInt(a, den);
+            dTdk -= _divDownInt((prices[i] * a) / den, den);
+            dPdkDivP += _divDownInt(prices[i], _mulDownInt(k, prices[i]) - a);
+            alphaDivPTn = (alphaDivPTn * b) / a;
```

Comment:
I have some doubts here.

The denominator according to the notion doc is `k * prices[i] / a - 1`, so if you say here that `den = k * prices[i] - a`, you're actually saying that you need to multiply the numerator by `a`.

This is reflected correctly when computing `T`.

But for `dTdK`...
Nvm got it as I was writing this :) you multiply the numerator by `a^2` and divide by `a`, so it's fine.